### DISCUSSION TITLE
exploding values in online competition

### DISCUSSION THREAD

**Discussion by rnz1234** [yasserfarouk/scml#69] [one-shot-qa] [answered]
Hi,
I see the last 2 days have exploding score values in the online tournaments (oneshot) - is it valid ? very different behavior than before

**Comment by yasserfarouk** [comment, marked as answer]
Yes we noticed that. 

The main reason is that the trading price of the intermediate product keeps going up at a rate of around 7% in some simulations. This is not a very large rate but when compound on 200 or 300 steps, it leads to this explosion.

The reason this was not happening before is that the system was incorrectly setting negotiation price range based on the catalog price instead of the trading price which was dampening this effect.

It seems that at least some agents do not have tests to check that they do not buy at a very unrealistic price. For seller agents, there is a hard limit on how unrealistic the price can be because it cannot be negative. For buyers, no such natural limit exist.

Notice that even a single buyer agent that systematically accepts unrealistically high prices can lead to this explosion because our market is both an oligopoly and an oligopsony. That is why we run the official competition on two rounds. Agents that fall victim to this problem will be removed in the qualification round and the problem should not happen in the finals.

We are investigating the issue thoroughly now and will update the answer when we have more details.

We completed our investigation. The cause of the issue was a bug that allowed agents to keep negotiating and losing after becoming bankrupt. The issue is resolved now and the tournaments with exploding results have been removed from the online competition.

It is still early to be sure but I think the problem was transitory and is resolved now.

**Comment by rnz1234** [comment]
isn't the price range (in ami.issues) fixed for the whole simulation ?

**Reply by yasserfarouk** [reply]
No. It is a function of the trading prices. See section 3.2 [here](http://www.yasserm.com/scml/scml2021oneshot.pdf)

**Reply by yasserfarouk** [reply]
@rnz1234 

We completed our investigation. The cause of the issue was a bug that allowed agents to keep negotiating and losing after becoming bankrupt. The issue is resolved now and the tournaments with exploding results have been removed from the online competition.

It is still early to be sure but I think the problem was transitory and is resolved now.

**Comment by rnz1234** [comment]
I have some issue with my latest loading which doesn't make sense:

[14, 'Traceback (most recent call last):\n File "/export/home/yasser/code/projects/negmas/negmas/situated/situated.py", line 3105, in call\n result = method(*args, **kwargs)\n File "/export/home/yasser/code/projects/negmas/negmas/situated/situated.py", line 1754, in on_negotiation_success_\n self.on_negotiation_success(contract, mechanism)\n File "/export/home/yasser/code/projects/scml/src/scml/oneshot/sysagents.py", line 46, in on_negotiation_success\n return self._obj.on_negotiation_success(contract, mechanism)\n File "/export/home/yasser/scmlrestores/_agents_oneshot/team_51/team_51/scml_agent/qlagent_extended_state.py", line 433, in on_negotiation_success\n if self._is_selling(mechanism):\nZeroDivisionError: float division by zero\n']

I don't divide there by anything, just checking that self.awi.current_exogenous_input_quantity or self.awi.current_exogenous_output_quantity are not smaller/equal to 0.

Why is it happening ?
it doesn't happen to me when I run in my env.

Can you check please if it is some system issue ?

thanks

**Comment by rnz1234** [comment]
I'm pretty sure there's some issue with simulation. I get call for the on_success() with exogenous input or output which 0 (and should not be - for example, input=0 for seller aka L0)

**Reply by rnz1234** [reply]
I've taken recent version as written in today's mail

**Reply by yasserfarouk** [reply]
Not sure I understand this issue. Do you mean that you get on_negotiation_success() and inside it you check the exogenous input or output and find it zero? If that is the case, there is nothing wrong. The exogenous input/output quantity is only guaranteed to be less than or equal to the number of lines but they can both be zero in some step. In these steps you should not really accept anything but of course you can.

**Reply by yasserfarouk** [reply]
I think we understand your issue now. You were doing a division by the exogenous quantity which can be zero. The way python tracebacks work is that the system gets the exception then it **goes back to the source code** to print the line content at which the error occurred. 

What happened is that a version of the agent was raising this exception, but you submitted another version later with changed source code **while the tournament is still running**. When python went back to your code, it reported whatever was at the line $433$ which no longer contains the division.

I can see that the issue is resolved in the latest version of your agent.

If you are satisfied with the answer, please delete this discussion about divisionByZero and exogenous quantity (starting from your second comment) as it is not relevant to the exploding values topic. Thanks
